internal/cli/cmd/cluster/grpcproxy.go

```diff
@@ -0,0 +1,199 @@
+
+func serveGRPCProxy(listener net.Listener, connect func(context.Context) (net.Conn, error)) error {
+	p, err := newGrpcProxy(connect)
+	if err != nil {
+		return err
+	}
+
+	return p.Serve(listener)
+}
+
+type grpcProxy struct {
+	mu sync.Mutex
+	*grpc.Server
+	backendClient *grpc.ClientConn
+	connect       func(context.Context) (net.Conn, error)
+}
+
+func newGrpcProxy(connect func(context.Context) (net.Conn, error)) (*grpcProxy, error) {
+	g := &grpcProxy{
+		connect: connect,
+	}
+
+	s := grpc.NewServer(grpc.UnknownServiceHandler(g.handler))
+	g.Server = s
+	return g, nil
+}
+
+func (g *grpcProxy) newBackendClient(ctx context.Context) (*grpc.ClientConn, error) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.backendClient != nil {
+		if g.backendClient.GetState() == connectivity.Ready {
+			fmt.Fprintf(console.Debug(context.Background()), "reused grpc connection\n")
+			return g.backendClient, nil
+		}
+		fmt.Fprintf(console.Debug(context.Background()), "cached grpc connection invalidated\n")
+		g.backendClient = nil
+	}
+
+	client, err := grpc.Dial("",
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:    time.Second * 30,
+			Timeout: time.Minute,
+		}),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(func(context.Context, string) (net.Conn, error) {
+			return g.connect(ctx)
+		}))
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Fprintf(console.Debug(context.Background()), "created new grpc connection\n")
+
+	g.backendClient = client
+	return client, nil
+}
+
+func (g *grpcProxy) handler(srv interface{}, serverStream grpc.ServerStream) error {
+	fullMethodName, ok := grpc.MethodFromServerStream(serverStream)
+	if !ok {
+		err := status.Errorf(codes.Internal, "reading method failed")
+		fmt.Fprintf(console.Errors(context.Background()), "reading method failed: %v\n", err)
+		return err
+	}
+	fmt.Fprintf(console.Debug(context.Background()), "handler %s\n", fullMethodName)
+
+	md, _ := metadata.FromIncomingContext(serverStream.Context())
+	outgoingCtx := metadata.NewOutgoingContext(serverStream.Context(), md.Copy())
+	backendConn, err := g.newBackendClient(outgoingCtx)
+	if err != nil {
+		fmt.Fprintf(console.Errors(context.Background()), "creating backend connection failed: %v\n", err)
+		return status.Errorf(codes.Internal, "failed connect to backend: %v", err)
+	}
+
+	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
+	defer clientCancel()
+
+	clientStream, err := grpc.NewClientStream(clientCtx, clientStreamDescForProxying, backendConn, fullMethodName)
+	if err != nil {
+		fmt.Fprintf(console.Errors(context.Background()), "failed to create client stream: %v\n", err)
+		return status.Errorf(codes.Internal, "failed create client: %v", err)
+	}
+
+	s2cErrChan := g.proxyServerToClient(serverStream, clientStream)
+	c2sErrChan := g.proxyClientToServer(clientStream, serverStream)
+	// Make sure to close both client and server connections
+	for i := 0; i < 2; i++ {
+		select {
```

Comment:
Should there be a case `<-serverStream.Context().Done()` for a clean cancellation?